src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrProviderStateFactory.java

```diff
@@ -111,7 +111,20 @@ private static Bundle extractCallingBundle(@NotNull Map<String, Object> authenti
                     } else {
                         final Object subService = authenticationInfo.get(ResourceResolverFactory.SUBSERVICE);
                         final String subServiceName = subService instanceof String ? (String) subService : null;
-                        session = repo.loginService(subServiceName, null);
+                        // let's shortcut the impersonation for service users, if impersonation was requested
```

Comment:
the comment confuses me... what you are handling here is getServiceResolver request that in addition comes with a hint that an impersonation should be performed.... i.e. calling impersonateFromService instead of just loginService.
the first part 'impersonation for service users is the one that confuses me.... impersonateFromService means that a service should perform an impersonation of any user. from the osgi service (bundle/subservice) a given specified user should be impersonated. it's not that the service user is in charge here... it could be but that's an impl detail.

Reply:
It is a shortcut, since impersonation would have been handled in https://github.com/apache/sling-org-apache-sling-jcr-resource/blob/f82678ae650a67538d261bcdc5d1e254039e9a37/src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrProviderStateFactory.java#L150, through calling https://github.com/apache/sling-org-apache-sling-jcr-resource/blob/f82678ae650a67538d261bcdc5d1e254039e9a37/src/main/java/org/apache/sling/jcr/resource/internal/helper/jcr/JcrProviderStateFactory.java#L189.

I think I should rephrase the comment to "[...] impersonation for services [...]", since that was my intention.

Reply:
👍

Reply:
This should be now clear with the changes from https://github.com/apache/sling-org-apache-sling-jcr-resource/pull/43/commits/b54140babc116cad4cefce98c7c57670145a18bc.